File: IOS/IOS/Mail/MailFilterLedger.swift
import Foundation

/// Persists which filter applications ran for the current rules generation (same idea as web `localStorage`).
enum MailFilterLedger {
    private static let storageVersion = 1

    private struct Entry: Codable, Sendable {
        var ruleId: String
        var actionFingerprint: String
        var rulesHash: String
    }

    private struct Store: Codable, Sendable {
        var version: Int
        var rulesHash: String
        var entries: [String: Entry]
    }

    private static func key(sessionEmail: String) -> String {
        let norm = KeyVaultKeychain.normalizeEmail(sessionEmail)
        return "elvish_mail_filter_ledger_v\(storageVersion)_\(norm)"
    }

    private static func read(sessionEmail: String) -> Store {
        let d = UserDefaults.standard.data(forKey: key(sessionEmail: sessionEmail)) ?? Data()
        guard !d.isEmpty,
              let s = try? JSONDecoder().decode(Store.self, from: d),
              s.version == storageVersion
        else {
            return Store(version: storageVersion, rulesHash: "", entries: [:])
        }
        return s
    }

    private static func write(_ store: Store, sessionEmail: String) {
        if let data = try? JSONEncoder().encode(store) {
            UserDefaults.standard.set(data, forKey: key(sessionEmail: sessionEmail))
        }
    }

    static func syncRulesHash(_ rulesHash: String, sessionEmail: String) {
        var s = read(sessionEmail: sessionEmail)
        if s.rulesHash != rulesHash {
            s.rulesHash = rulesHash
            s.entries = [:]
            write(s, sessionEmail: sessionEmail)
        }
    }

    static func alreadyApplied(messageId: String, ruleId: String, actionFingerprint: String, rulesHash: String, sessionEmail: String) -> Bool {
        guard !messageId.isEmpty, !ruleId.isEmpty, !rulesHash.isEmpty else { return false }
        let s = read(sessionEmail: sessionEmail)
        guard s.rulesHash == rulesHash, let e = s.entries[messageId] else { return false }
        return e.ruleId == ruleId && e.actionFingerprint == actionFingerprint && e.rulesHash == rulesHash
    }

    static func record(messageId: String, ruleId: String, actionFingerprint: String, rulesHash: String, sessionEmail: String) {
        guard !messageId.isEmpty, !ruleId.isEmpty, !rulesHash.isEmpty else { return }
        var s = read(sessionEmail: sessionEmail)
        if s.rulesHash != rulesHash {
            s.rulesHash = rulesHash
            s.entries = [:]
        }
        s.entries[messageId] = Entry(ruleId: ruleId, actionFingerprint: actionFingerprint, rulesHash: rulesHash)
        write(s, sessionEmail: sessionEmail)
    }

    static func pruneMissing(messageIds: Set<String>, sessionEmail: String) {
        var s = read(sessionEmail: sessionEmail)
        s.entries = s.entries.filter { messageIds.contains($0.key) }
        write(s, sessionEmail: sessionEmail)
    }

    static func clear(sessionEmail: String) {
        UserDefaults.standard.removeObject(forKey: key(sessionEmail: sessionEmail))
    }
}

File: flutter/elvish_mail/lib/mail/mail_filter_ledger.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

import '../keyvault/secure_account_store.dart';

/// Persists which filter applications ran for the current rules generation (web `localStorage`).
class MailFilterLedger {
  MailFilterLedger._();

  static const int storageVersion = 1;

  static String _key(String sessionEmail) {
    final norm = SecureAccountStore.normalizeEmail(sessionEmail);
    return 'elvish_mail_filter_ledger_v${storageVersion}_$norm';
  }

  static Future<_Store> _read(String sessionEmail) async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_key(sessionEmail));
    if (raw == null || raw.isEmpty) {
      return _Store(version: storageVersion, rulesHash: '', entries: {});
    }
    try {
      final j = jsonDecode(raw) as Map<String, dynamic>;
      if ((j['version'] as num?)?.toInt() != storageVersion) {
        return _Store(version: storageVersion, rulesHash: '', entries: {});
      }
      final entries = <String, _Entry>{};
      final rawE = j['entries'];
      if (rawE is Map) {
        for (final e in rawE.entries) {
          if (e.value is Map<String, dynamic>) {
            final m = e.value as Map<String, dynamic>;
            entries[e.key] = _Entry(
              ruleId: m['ruleId'] as String? ?? '',
              actionFingerprint: m['actionFingerprint'] as String? ?? '',
              rulesHash: m['rulesHash'] as String? ?? '',
            );
          }
        }
      }
      return _Store(
        version: storageVersion,
        rulesHash: j['rulesHash'] as String? ?? '',
        entries: entries,
      );
    } catch (_) {
      return _Store(version: storageVersion, rulesHash: '', entries: {});
    }
  }

  static Future<void> _write(_Store store, String sessionEmail) async {
    final prefs = await SharedPreferences.getInstance();
    final payload = jsonEncode({
      'version': store.version,
      'rulesHash': store.rulesHash,
      'entries': {
        for (final e in store.entries.entries)
          e.key: {
            'ruleId': e.value.ruleId,
            'actionFingerprint': e.value.actionFingerprint,
            'rulesHash': e.value.rulesHash,
          },
      },
    });
    await prefs.setString(_key(sessionEmail), payload);
  }

  static Future<void> syncRulesHash(String rulesHash, String sessionEmail) async {
    var s = await _read(sessionEmail);
    if (s.rulesHash != rulesHash) {
      s = _Store(version: storageVersion, rulesHash: rulesHash, entries: {});
      await _write(s, sessionEmail);
    }
  }

  static Future<bool> alreadyApplied({
    required String messageId,
    required String ruleId,
    required String actionFingerprint,
    required String rulesHash,
    required String sessionEmail,
  }) async {
    if (messageId.isEmpty || ruleId.isEmpty || rulesHash.isEmpty) {
      return false;
    }
    final s = await _read(sessionEmail);
    if (s.rulesHash != rulesHash) {
      return false;
    }
    final e = s.entries[messageId];
    if (e == null) {
      return false;
    }
    return e.ruleId == ruleId &&
        e.actionFingerprint == actionFingerprint &&
        e.rulesHash == rulesHash;
  }

  static Future<void> record({
    required String messageId,
    required String ruleId,
    required String actionFingerprint,
    required String rulesHash,
    required String sessionEmail,
  }) async {
    if (messageId.isEmpty || ruleId.isEmpty || rulesHash.isEmpty) {
      return;
    }
    var s = await _read(sessionEmail);
    if (s.rulesHash != rulesHash) {
      s = _Store(version: storageVersion, rulesHash: rulesHash, entries: {});
    }
    s.entries[messageId] = _Entry(
      ruleId: ruleId,
      actionFingerprint: actionFingerprint,
      rulesHash: rulesHash,
    );
    await _write(s, sessionEmail);
  }

  static Future<void> pruneMissing(Set<String> messageIds, String sessionEmail) async {
    var s = await _read(sessionEmail);
    s.entries.removeWhere((id, _) => !messageIds.contains(id));
    await _write(s, sessionEmail);
  }

  static Future<void> clear(String sessionEmail) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_key(sessionEmail));
  }
}

class _Entry {
  _Entry({
    required this.ruleId,
    required this.actionFingerprint,
    required this.rulesHash,
  });

  final String ruleId;
  final String actionFingerprint;
  final String rulesHash;
}

class _Store {
  _Store({
    required this.version,
    required this.rulesHash,
    required this.entries,
  });

  final int version;
  final String rulesHash;
  final Map<String, _Entry> entries;
}
